Mathlib/Tactic/FieldSimp/Lemmas.lean

```diff
@@ -0,0 +1,549 @@
+    (h : l₁.eval / l₂.eval = l.eval) :
+    ((r₁, x) ::ᵣ l₁).eval / ((r₂, x) ::ᵣ l₂).eval = ((r₁ - r₂, x) ::ᵣ l).eval := by
+  simp only [← h, eval_cons, div_eq_mul_inv, mul_inv, ← zpow'_neg, mul_assoc]
+  congr! 1
+  rw [mul_comm, mul_assoc]
+  nth_rewrite 2 [mul_comm]
+  rw [← zpow'_add, ← sub_eq_add_neg]
+
+theorem div_eq_eval₃ [CommGroupWithZero M] {a₁ : ℤ × M} (a₂ : ℤ × M) {l₁ l₂ l : NF M}
+    (h : (a₁ ::ᵣ l₁).eval / l₂.eval = l.eval) :
+    (a₁ ::ᵣ l₁).eval / (a₂ ::ᵣ l₂).eval = ((-a₂.1, a₂.2) ::ᵣ l).eval := by
+  simp only [eval_cons, ← h, zpow'_neg, div_eq_mul_inv, mul_inv, mul_assoc]
+
+theorem div_eq_eval [GroupWithZero M] {l₁ l₂ l : NF M} {x₁ x₂ : M} (hx₁ : x₁ = l₁.eval)
+    (hx₂ : x₂ = l₂.eval) (h : l₁.eval / l₂.eval = l.eval) :
+    x₁ / x₂ = l.eval := by
+  rw [hx₁, hx₂, h]
+
+theorem eval_mul_eval_cons [GroupWithZero M] (n : ℤ) (e : M) {L l l' : NF M}
+    (h : L.eval * l.eval = l'.eval) :
+    L.eval * ((n, e) ::ᵣ l).eval = ((n, e) ::ᵣ l').eval := by
+  rw [eval_cons, eval_cons, ← h, mul_assoc]
+
+theorem eval_mul_eval_cons_zero [GroupWithZero M] {e : M} {L l l' l₀ : NF M}
+    (h : L.eval * l.eval = l'.eval) (h' : ((0, e) ::ᵣ l).eval = l₀.eval) :
+    L.eval * l₀.eval = ((0, e) ::ᵣ l').eval := by
+  rw [← eval_mul_eval_cons 0 e h, h']
+
+theorem eval_cons_mul_eval [CommGroupWithZero M] (n : ℤ) (e : M) {L l l' : NF M}
+    (h : L.eval * l.eval = l'.eval) :
+    ((n, e) ::ᵣ L).eval * l.eval = ((n, e) ::ᵣ l').eval := by
+  rw [eval_cons, eval_cons, ← h, mul_assoc, mul_assoc]
+  congr! 1
+  rw [mul_comm]
+
+theorem eval_cons_mul_eval_cons_neg [CommGroupWithZero M] (n : ℤ) {e : M} (he : e ≠ 0)
+    {L l l' : NF M} (h : L.eval * l.eval = l'.eval) :
+    ((n, e) ::ᵣ L).eval * ((-n, e) ::ᵣ l).eval = l'.eval := by
+  rw [mul_eq_eval₂ n (-n) e h]
+  simp [zpow'_zero_of_ne_zero he]
+
+theorem cons_eq_div_of_eq_div [CommGroupWithZero M] (n : ℤ) (e : M) {t t_n t_d : NF M}
+    (h : t.eval = t_n.eval / t_d.eval) :
+    ((n, e) ::ᵣ t).eval = ((n, e) ::ᵣ t_n).eval / t_d.eval := by
+  simp only [eval_cons, h, div_eq_mul_inv]
+  rw [mul_comm, ← mul_assoc, mul_comm _ t_n.eval]
+
+theorem cons_eq_div_of_eq_div' [CommGroupWithZero M] (n : ℤ) (e : M) {t t_n t_d : NF M}
+    (h : t.eval = t_n.eval / t_d.eval) :
+    ((-n, e) ::ᵣ t).eval = t_n.eval / ((n, e) ::ᵣ t_d).eval := by
+  simp only [eval_cons, h, zpow'_neg, div_eq_mul_inv, mul_inv]
+  rw [← mul_assoc, mul_comm]
+
+theorem cons_zero_eq_div_of_eq_div [CommGroupWithZero M] (e : M) {t t_n t_d : NF M}
+    (h : t.eval = t_n.eval / t_d.eval) :
+    ((0, e) ::ᵣ t).eval = ((1, e) ::ᵣ t_n).eval / ((1, e) ::ᵣ t_d).eval := by
+  simp only [eval_cons, h, div_eq_mul_inv, mul_inv, mul_assoc]
+  congr! 1
+  rw [← mul_assoc, mul_comm (zpow' e 1), mul_assoc]
+  congr! 1
+  rw [← zpow'_neg, ← zpow'_add]
+  congr
+
+instance : Inv (NF M) where
+  inv l := l.map fun (a, x) ↦ (-a, x)
+
+-- generalize library `List.prod_inv`
+theorem _root_.List.prod_inv₀ {K : Type*} [DivisionCommMonoid K] :
```

Comment:
Move out to the library?

Reply:
Moved. I'm not sure what the "generalize" part means --- this would only make sense if the definition of `prod'` is changed.